### PR TITLE
Add replication from Cozy :inbox_tray:

### PR DESCRIPTION
**Pull request by kosssi**

### CODE REVIEW

File: src/offline.js [hunk 1]
```diff
@@ -6,7 +6,26 @@ export function init (cozy, { options = {}, docTypes = [] }) {
...
+      'change', 'paused', 'active', 'denied', 'complete', 'error'
+    ]
+    for (let eventName of eventNames) {
+      if (typeof events[eventName] === 'function') {
+        replication.on(eventName, events[eventName])
+      }
+    }
+    return replication
+  } else {
+    throw new Error(`You could add this docType: ${docType} to offline.`)
```
Comment (aenario) [inline src/offline.js]: s/could/should/

File: src/offline.js [hunk 2]
```diff
@@ -6,7 +6,26 @@ export function init (cozy, { options = {}, docTypes = [] }) {
...
+
+export function replicateFromCozy (cozy, docType, options = {}, events = {}) {
+  if (cozy._offline_databases && docType in cozy._offline_databases) {
+    const url = cozy._url + '/data/' + docType
+    let db = cozy._offline_databases[docType]
+    let replication = db.replicate.from(url, options)
+    const eventNames = [
+      'change', 'paused', 'active', 'denied', 'complete', 'error'
+    ]
+    for (let eventName of eventNames) {
```
Comment (aenario) [inline src/offline.js]: if you return the replication, I am not sure we need the events: the using code can call .on(x) itself.

File: test/integration/offline.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      for (var i = 0, l = docs.length; i < l; i++) {
+        await cozy.delete(DOCTYPE, docs[i])
+      }
+    }
+  })
+
+  it('can replicate database from cozy', async function () {
+    cozy.offline.addDocType(DOCTYPE, {adapter: 'memory'})
+    let complete = await cozy.offline.replicateFromCozy(DOCTYPE)
+    complete.docs_written.should.not.equal(67)
```
Comment (aenario) [inline test/integration/offline.js]: ?

**Comment by kosssi** [comment]
it's done @aenario ;)

File: src/offline.js (hunk 2 above)
Comment (kosssi) [inline src/offline.js]: it's for not touch replication and simplify it.

File: src/offline.js [hunk 4]
```diff
@@ -1,12 +1,53 @@
...
+
+export function hasDatabase (cozy, doctype) {
+  return cozy._offline_databases && doctype in cozy._offline_databases
+}
+
+export function getDatabase (cozy, doctype) {
+  if (hasDatabase(cozy, doctype)) {
+    return cozy._offline_databases[doctype]
+  }
+  return false
```
Comment (enguerran) [inline src/offline.js]: Why `getDatabase` returns a pouchDB instance (`Object`) or `false`. I know JavaScript allows weak typing, but shouldn't we return consistant JavaScript types like `Object || undefined`.

File: test/integration/offline.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    })
+    for (var i = 0, l = docs.length; i < l; i++) {
+      docs[i] = await cozy.create(DOCTYPE, docs[i])
+    }
+  })
+
+  after(async function () {
+    if (COZY_STACK_VERSION === '3') {
+      for (var i = 0, l = docs.length; i < l; i++) {
+        await cozy.delete(DOCTYPE, docs[i])
```
Comment (enguerran) [inline test/integration/offline.js]: I am curious about this, what the meaning of `l`?
Why don't use:
```javascript
docs.forEach(function(doc) { await cozy.delete(DOCTYPE, doc) })
```

File: test/integration/offline.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  before(async function () {
+    if (COZY_STACK_VERSION === '2') {
+      return this.skip()
+    }
+    cozy = new Cozy({
+      cozyURL: COZY_STACK_URL
+    })
+    for (var i = 0, l = docs.length; i < l; i++) {
+      docs[i] = await cozy.create(DOCTYPE, docs[i])
```
Comment (enguerran) [inline test/integration/offline.js]: same remark as [below](https://github.com/cozy/cozy-client-js/pull/41/files#r97268831)

```javascript
docs = docs.map(function(doc) { return await cozy.create(DOCTYPE, doc) })
```

File: test/integration/offline.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+  after(async function () {
+    if (COZY_STACK_VERSION === '3') {
+      for (var i = 0, l = docs.length; i < l; i++) {
+        await cozy.delete(DOCTYPE, docs[i])
+      }
+    }
+  })
+
+  it('can replicate database from cozy', async function () {
```
Comment (enguerran) [inline test/integration/offline.js]: I cannot understand the intention beyond the 
```javascript
let complete = await cozy.offline.replicateFromCozy(DOCTYPE)
complete.docs_written.should.not.equal(0)
complete = await cozy.offline.replicateFromCozy(DOCTYPE)
complete.docs_written.should.equal(0)
```

File: test/integration/offline.js (hunk 6 above)
Comment (aenario) [inline test/integration/offline.js]: Does not work with await, you'd have to use Promise.all, but i find it less readable

File: test/integration/offline.js (hunk 7 above)
Comment (aenario) [inline test/integration/offline.js]: First replication does something, second does not (as the two databases are already sync)

File: test/integration/offline.js (hunk 5 above)
Comment (kosssi) [inline test/integration/offline.js]: `l` is for lenght ;) I just copy the code from https://github.com/cozy/cozy-client-js/blob/master/test/integration/mango.js

Comment (kosssi) [inline test/integration/offline.js]: forEach is not compatible with await.

**Comment by kosssi** [comment]
I think it ok ;)

File: test/integration/offline.js (hunk 5 above)
Comment (enguerran) [inline test/integration/offline.js]: regarding [@aenario 's comment](https://github.com/cozy/cozy-client-js/pull/41#discussion_r97283316), it is not.

Why do you instantiate two variables in the for loop? I guess `for (var i = 0 ; i < docs.length ; i++)` should work the same as `for (var i = 0, l = docs.length ; i < l ; i++)`. Or do I misunderstand something?

cf. https://jsperf.com/array-prototype-length/1 ;)

File: test/integration/offline.js (hunk 7 above)
Comment (enguerran) [inline test/integration/offline.js]: okay

File: test/integration/offline.js (hunk 5 above)
Comment (aenario) [inline test/integration/offline.js]: Yep, it is a stupid reflex I got from somewhere, which probably got it from the time before JS runtimes optimised it. Good on kossi to follow the repo codestyle, but we can change it as I agree it is less readable.

**Pull request merged**